File: f8/hamming.java
class Hamming {
  public static void main(String[] args) {
    System.out.println(Hamming.distance(5,5));
    System.out.println(Hamming.distance(24,16));
    System.out.println(Hamming.distance(1,3));
    System.out.println(Hamming.distance(1,2));
  }

  public static int distance(int a, int b) {
    int diff = a ^ b;
    return Hamming.weight(diff);
  }

  public static int weight(int a) {
    int count = 0;
    int pow = 1;
    for (int i = 0; i < 32; i++) {
      if ((a & pow) == pow) {
        count++;
      }
      pow = pow * 2;
    }
    return count;
  }
}

File: f8/hamming.rb
class Hamming
  def self.distance a, b
    diff = a ^ b
    p Hamming.weight(diff)
  end

  def self.weight a
    bit = 2 ** 0
    count = 0
    (0..32).each do |i|
      if (bit & a) == bit
        count +=1
      end
      bit *= 2
    end
    count
  end
end

Hamming.distance(3,3 )
Hamming.distance(16,3)
Hamming.distance(24,8)
Hamming.distance(7,15)

RSpec.describe Hamming do
  it 'should run' do
    expect(1).to eq(1)
  end
end
